src/cupie.c

```diff
@@ -0,0 +1,879 @@
+        }
+        arg0->base.counter = 0x3c;
+        rand = Rand16();
+        mask = 1;
+        if (rand & mask) {
+            arg0->unk83 = 1;
+            arg0->unk78 = sub_080A3A9C;
+            arg0->base.flags &= ~2;
+            arg0->base.flags |= 0x140;
+            arg0->base.flags |= 0x800;
+        }
+    }
+}
+
+void sub_080A31D8(struct Object2* arg0) {
+    s32 r4, r3;
+    arg0->base.flags |= 4;
+    if (arg0->base.flags & 2) {
+        arg0->kirby3 = sub_0803D368(arg0);
+    }
+    if (arg0->base.x > arg0->kirby3->base.base.x) {
+        if (++arg0->unk9E > 0xf) {
+            arg0->kirby3 = sub_0803D368(arg0);
+            arg0->unk9E = 0;
+            arg0->unkA0 = arg0->kirby3->base.base.x >> 8;
+            arg0->unkA2 = arg0->kirby3->base.base.y >> 8;
+        }
+        r4 = arg0->base.y & 0xFFFFF000;
+        r3 = ((arg0->unkA2 - 0x40) << 8) & 0xFFFFF000;
+        // Stalk Y position
+        if (r4 > r3) {
+            arg0->base.yspeed += 8;
+            if (arg0->base.yspeed > 0x100) {
+                arg0->base.yspeed = 0x100;
+            }
+        }
+        else if (r4 < r3) {
+            if ((arg0->base.yspeed -= 8) < -0x100) {
+                arg0->base.yspeed = -0x100;
+            }
+        }
+        else {
+            if (arg0->base.yspeed < 0) {
+                arg0->base.yspeed += 4;
+                if (arg0->base.yspeed > 0) {
+                    arg0->base.yspeed = 0;
+                }
+            }
+            else {
+                arg0->base.yspeed -= 4;
+                if (arg0->base.yspeed < 0) {
+                    arg0->base.yspeed = 0;
+                }
+            }
+        }
+        r4 = arg0->base.x & 0xFFFFF000;
+        r3 = ((arg0->unkA0 + 0x60) << 8) & 0xFFFFF000;
+        // Stalk X position
+        if (r4 < r3) {
+            arg0->base.flags &= ~1;
+            arg0->base.xspeed += 8;
+            if (arg0->base.xspeed > 0x100) {
+                arg0->base.xspeed = 0x100;
+            }
+            else if (arg0->base.xspeed < -0x100) {
+                arg0->base.xspeed = -0x100;
+            }
+        }
+        else if (r4 > r3) {
+            arg0->base.flags |= 1;
+            if (arg0->base.flags & 1) {
+                arg0->base.xspeed -= 8;
+                if (arg0->base.xspeed < -0x100) {
+                    arg0->base.xspeed = -0x100;
+                }
+                else if (arg0->base.xspeed > 0x100) {
+                    arg0->base.xspeed = 0x100;
+                }
+            }
+            else {
+                arg0->base.xspeed += 8;
+                if (arg0->base.xspeed > 0x100) {
+                    arg0->base.xspeed = 0x100;
+                }
+                else if (arg0->base.xspeed < -0x100) {
+                    arg0->base.xspeed = -0x100;
+                }
+            }
+        }
+        else {
+            if (arg0->base.xspeed < 0) {
+                if ((arg0->base.xspeed += 4) << 0x10 > 0) {
```

Comment:
Add a comment here since this is a compiler trick that can invoke defined behavior